File: src/core/xml.ts
import type { SourceRange } from './model';

export interface XmlAttribute {
  name: string;
  value: string;
  range: SourceRange;
}

export interface XmlNode {
  name: string;
  openStart: number;
  openEnd: number;
  closeStart: number;
  attributes: XmlAttribute[];
  children: XmlNode[];
}

function localName(name: string): string {
  return name.slice(name.lastIndexOf(':') + 1);
}

export function decodeXml(value: string): string {
  return value.replace(/&(#x[\da-f]+|#\d+|amp|lt|gt|quot|apos);/gi, (entity, body: string) => {
    const lower = body.toLowerCase();
    if (lower === 'amp') return '&';
    if (lower === 'lt') return '<';
    if (lower === 'gt') return '>';
    if (lower === 'quot') return '"';
    if (lower === 'apos') return "'";
    const radix = lower.startsWith('#x') ? 16 : 10;
    const digits = lower.slice(radix === 16 ? 2 : 1);
    const code = Number.parseInt(digits, radix);
    return Number.isInteger(code) && code >= 0 && code <= 0x10FFFF
      ? String.fromCodePoint(code)
      : entity;
  });
}

function attributesIn(text: string, start: number, end: number): XmlAttribute[] {
  const attributes: XmlAttribute[] = [];
  let cursor = start;
  while (cursor < end) {
    while (cursor < end && /\s/.test(text[cursor]!)) cursor++;
    if (cursor >= end || text[cursor] === '/') break;
    const nameStart = cursor;
    while (cursor < end && !/[\s=/]/.test(text[cursor]!)) cursor++;
    const rawName = text.slice(nameStart, cursor);
    while (cursor < end && /\s/.test(text[cursor]!)) cursor++;
    if (!rawName || text[cursor] !== '=') {
      while (cursor < end && !/\s/.test(text[cursor]!) && text[cursor] !== '/') cursor++;
      continue;
    }
    cursor++;
    while (cursor < end && /\s/.test(text[cursor]!)) cursor++;
    const quote = text[cursor];
    if (quote !== '"' && quote !== "'") continue;
    cursor++;
    const valueStart = cursor;
    while (cursor < end && text[cursor] !== quote) cursor++;
    if (cursor >= end) break;
    const rawValue = text.slice(valueStart, cursor);
    const leading = rawValue.length - rawValue.trimStart().length;
    const trimmed = rawValue.trim();
    attributes.push({
      name: localName(rawName),
      value: decodeXml(trimmed),
      range: { offset: valueStart + leading, length: trimmed.length },
    });
    cursor++;
  }
  return attributes;
}

/** Small, tolerant XML reader used by static manifest adapters. */
export function parseXml(text: string): XmlNode[] {
  const roots: XmlNode[] = [];
  const stack: XmlNode[] = [];
  let cursor = 0;
  while (cursor < text.length) {
    const open = text.indexOf('<', cursor);
    if (open < 0) break;
    if (text.startsWith('<!--', open)) {
      const end = text.indexOf('-->', open + 4);
      cursor = end < 0 ? text.length : end + 3;
      continue;
    }
    if (text.startsWith('<![CDATA[', open)) {
      const end = text.indexOf(']]>', open + 9);
      cursor = end < 0 ? text.length : end + 3;
      continue;
    }
    if (text.startsWith('<?', open)) {
      const end = text.indexOf('?>', open + 2);
      cursor = end < 0 ? text.length : end + 2;
      continue;
    }
    if (text.startsWith('<!', open)) {
      const end = text.indexOf('>', open + 2);
      cursor = end < 0 ? text.length : end + 1;
      continue;
    }

    let quote: '"' | "'" | undefined;
    let end = open + 1;
    for (; end < text.length; end++) {
      const char = text[end]!;
      if (quote) {
        if (char === quote) quote = undefined;
      } else if (char === '"' || char === "'") quote = char;
      else if (char === '>') break;
    }
    if (end >= text.length) break;

    let contentStart = open + 1;
    while (contentStart < end && /\s/.test(text[contentStart]!)) contentStart++;
    if (text[contentStart] === '/') {
      contentStart++;
      while (contentStart < end && /\s/.test(text[contentStart]!)) contentStart++;
      let nameEnd = contentStart;
      while (nameEnd < end && !/\s/.test(text[nameEnd]!)) nameEnd++;
      const name = localName(text.slice(contentStart, nameEnd));
      for (let i = stack.length - 1; i >= 0; i--) {
        if (stack[i]!.name !== name) continue;
        stack[i]!.closeStart = open;
        stack.length = i;
        break;
      }
    } else {
      let nameEnd = contentStart;
      while (nameEnd < end && !/[\s/]/.test(text[nameEnd]!)) nameEnd++;
      const rawName = text.slice(contentStart, nameEnd);
      if (rawName !== '') {
        const selfClosing = /\/\s*$/.test(text.slice(nameEnd, end));
        const node: XmlNode = {
          name: localName(rawName),
          openStart: open,
          openEnd: end + 1,
          closeStart: end + 1,
          attributes: attributesIn(text, nameEnd, end),
          children: [],
        };
        const parent = stack[stack.length - 1];
        if (parent) parent.children.push(node);
        else roots.push(node);
        if (!selfClosing) stack.push(node);
      }
    }
    cursor = end + 1;
  }
  for (const node of stack) node.closeStart = text.length;
  return roots;
}

export function xmlChild(node: XmlNode, name: string): XmlNode | undefined {
  return node.children.find((candidate) => candidate.name === name);
}

export function xmlAttribute(node: XmlNode, name: string): XmlAttribute | undefined {
  return node.attributes.find((candidate) => candidate.name === name);
}

export function xmlTextValue(
  text: string,
  node: XmlNode | undefined,
): { value: string; range: SourceRange } | undefined {
  if (!node) return undefined;
  const raw = text.slice(node.openEnd, node.closeStart);
  const plain = raw.replace(/<!--[^]*?-->/g, '').replace(/<!\[CDATA\[([^]*?)]]>/g, '$1');
  const trimmed = plain.trim();
  const value = decodeXml(trimmed);
  if (value === '') return undefined;
  const rawValueIndex = raw.indexOf(trimmed);
  const offset = rawValueIndex < 0 ? node.openEnd : node.openEnd + rawValueIndex;
  return { value, range: { offset, length: trimmed.length } };
}

File: jetbrains/src/main/kotlin/dev/pacmon/jetbrains/core/Xml.kt
package dev.pacmon.jetbrains.core

internal data class XmlAttribute(
    val name: String,
    val value: String,
    val range: SourceRange,
)

internal data class XmlNode(
    val name: String,
    val openStart: Int,
    val openEnd: Int,
    var closeStart: Int,
    val attributes: List<XmlAttribute>,
    val children: MutableList<XmlNode> = mutableListOf(),
)

private fun localXmlName(name: String): String = name.substringAfterLast(':')

internal fun decodeXml(value: String): String =
    Regex("&(#x[\\da-f]+|#\\d+|amp|lt|gt|quot|apos);", RegexOption.IGNORE_CASE).replace(value) { match ->
        val body = match.groupValues[1].lowercase()
        when (body) {
            "amp" -> "&"
            "lt" -> "<"
            "gt" -> ">"
            "quot" -> "\""
            "apos" -> "'"
            else -> {
                val radix = if (body.startsWith("#x")) 16 else 10
                val digits = body.drop(if (radix == 16) 2 else 1)
                val codePoint = digits.toIntOrNull(radix)
                if (codePoint == null || !Character.isValidCodePoint(codePoint)) {
                    match.value
                } else {
                    runCatching { Character.toChars(codePoint).concatToString() }.getOrDefault(match.value)
                }
            }
        }
    }

private fun attributesIn(text: String, start: Int, end: Int): List<XmlAttribute> {
    val attributes = mutableListOf<XmlAttribute>()
    var cursor = start
    while (cursor < end) {
        while (cursor < end && text[cursor].isWhitespace()) cursor++
        if (cursor >= end || text[cursor] == '/') break
        val nameStart = cursor
        while (cursor < end && !text[cursor].isWhitespace() && text[cursor] !in setOf('=', '/')) cursor++
        val rawName = text.substring(nameStart, cursor)
        while (cursor < end && text[cursor].isWhitespace()) cursor++
        if (rawName.isEmpty() || cursor >= end || text[cursor] != '=') {
            while (cursor < end && !text[cursor].isWhitespace() && text[cursor] != '/') cursor++
            continue
        }
        cursor++
        while (cursor < end && text[cursor].isWhitespace()) cursor++
        if (cursor >= end) break
        val quote = text[cursor]
        if (quote != '\'' && quote != '"') continue
        cursor++
        val valueStart = cursor
        while (cursor < end && text[cursor] != quote) cursor++
        if (cursor >= end) break
        val rawValue = text.substring(valueStart, cursor)
        val leading = rawValue.length - rawValue.trimStart().length
        val trimmed = rawValue.trim()
        attributes.add(
            XmlAttribute(
                localXmlName(rawName),
                decodeXml(trimmed),
                SourceRange(valueStart + leading, trimmed.length),
            ),
        )
        cursor++
    }
    return attributes
}

/** Small, tolerant XML reader used by static manifest adapters. */
internal fun parseXml(text: String): List<XmlNode> {
    val roots = mutableListOf<XmlNode>()
    val stack = mutableListOf<XmlNode>()
    var cursor = 0
    while (cursor < text.length) {
        val open = text.indexOf('<', cursor)
        if (open < 0) break
        val marker = when {
            text.startsWith("<!--", open) -> "-->" to 4
            text.startsWith("<![CDATA[", open) -> "]]>" to 9
            text.startsWith("<?", open) -> "?>" to 2
            text.startsWith("<!", open) -> ">" to 2
            else -> null
        }
        if (marker != null) {
            val found = text.indexOf(marker.first, open + marker.second)
            cursor = if (found < 0) text.length else found + marker.first.length
            continue
        }

        var quote: Char? = null
        var end = open + 1
        while (end < text.length) {
            val ch = text[end]
            if (quote != null) {
                if (ch == quote) quote = null
            } else if (ch == '\'' || ch == '"') {
                quote = ch
            } else if (ch == '>') {
                break
            }
            end++
        }
        if (end >= text.length) break

        var contentStart = open + 1
        while (contentStart < end && text[contentStart].isWhitespace()) contentStart++
        if (contentStart < end && text[contentStart] == '/') {
            contentStart++
            while (contentStart < end && text[contentStart].isWhitespace()) contentStart++
            var nameEnd = contentStart
            while (nameEnd < end && !text[nameEnd].isWhitespace()) nameEnd++
            val name = localXmlName(text.substring(contentStart, nameEnd))
            for (index in stack.indices.reversed()) {
                if (stack[index].name != name) continue
                stack[index].closeStart = open
                while (stack.size > index) stack.removeAt(stack.lastIndex)
                break
            }
        } else {
            var nameEnd = contentStart
            while (nameEnd < end && !text[nameEnd].isWhitespace() && text[nameEnd] != '/') nameEnd++
            val rawName = text.substring(contentStart, nameEnd)
            if (rawName.isNotEmpty()) {
                val selfClosing = Regex("/\\s*$").containsMatchIn(text.substring(nameEnd, end))
                val node = XmlNode(
                    localXmlName(rawName),
                    open,
                    end + 1,
                    end + 1,
                    attributesIn(text, nameEnd, end),
                )
                stack.lastOrNull()?.children?.add(node) ?: roots.add(node)
                if (!selfClosing) stack.add(node)
            }
        }
        cursor = end + 1
    }
    stack.forEach { it.closeStart = text.length }
    return roots
}

internal fun xmlChild(node: XmlNode, name: String): XmlNode? = node.children.firstOrNull { it.name == name }

internal fun xmlAttribute(node: XmlNode, name: String): XmlAttribute? =
    node.attributes.firstOrNull { it.name == name }

internal fun xmlTextValue(text: String, node: XmlNode?): Pair<String, SourceRange>? {
    node ?: return null
    val raw = text.substring(node.openEnd.coerceAtMost(text.length), node.closeStart.coerceAtMost(text.length))
    val plain = raw.replace(Regex("<!--[\\s\\S]*?-->"), "")
        .replace(Regex("<!\\[CDATA\\[([\\s\\S]*?)]]>")) { it.groupValues[1] }
    val trimmed = plain.trim()
    val value = decodeXml(trimmed)
    if (value.isEmpty()) return null
    val relative = raw.indexOf(trimmed).coerceAtLeast(0)
    return value to SourceRange(node.openEnd + relative, trimmed.length)
}
